src/main/scala/devices/i2c/I2C.scala

```diff
@@ -0,0 +1,515 @@
+import util._
+import regmapper._
+import uncore.tilelink2._
+import rocketchip.PeripheryBusConfig
+import util.AsyncResetRegVec
+import sifive.blocks.devices.gpio.{GPIOPinCtrl}
+
+case class I2CConfig(address: BigInt)
+
+trait HasI2CParameters {
+  implicit val p: Parameters
+  val params: I2CConfig
+  val c = params
+}
+
+class I2CPin extends Bundle {
+  val in  = Bool(INPUT)
+  val out = Bool(OUTPUT)
+  val oe  = Bool(OUTPUT)
+}
+
+class I2CPort extends Bundle {
+  val scl = new I2CPin
+  val sda = new I2CPin
+}
+
+trait I2CBundle extends Bundle with HasI2CParameters {
+  val port = new I2CPort
+}
+
+trait I2CModule extends Module with HasI2CParameters with HasRegMap {
+  val io: I2CBundle
+
+  val I2C_CMD_NOP   = UInt(0x00)
+  val I2C_CMD_START = UInt(0x01)
+  val I2C_CMD_STOP  = UInt(0x02)
+  val I2C_CMD_WRITE = UInt(0x04)
+  val I2C_CMD_READ  = UInt(0x08)
+
+  class PrescalerBundle extends Bundle{
+    val hi = UInt(8.W)
+    val lo = UInt(8.W)
+  }
+
+  class ControlBundle extends Bundle{
+    val coreEn             = Bool()
+    val intEn              = Bool()
+    val reserved           = UInt(6.W)
+  }
+
+  class CommandBundle extends Bundle{
+    val start              = Bool()
+    val stop               = Bool()
+    val read               = Bool()
+    val write              = Bool()
+    val ack                = Bool()
+    val reserved           = UInt(2.W)
+    val irqAck             = Bool()
+  }
+
+  class StatusBundle extends Bundle{
+    val receivedAck        = Bool()    // received aknowledge from slave
+    val busy               = Bool()
+    val arbLost            = Bool()
+    val reserved           = UInt(3.W)
+    val transferInProgress = Bool()
+    val irqFlag            = Bool()
+  }
+
+  // control state visible to SW/driver
+  val prescaler    = Reg(init = (new PrescalerBundle).fromBits(0xFFFF.U))
+  val control      = Reg(init = (new ControlBundle).fromBits(0.U))
+  val transmitData = Reg(init = UInt(0, 8.W))
+  val receivedData = Reg(init = UInt(0, 8.W))
+  val cmd          = Reg(init = (new CommandBundle).fromBits(0.U))
+  val status       = Reg(init = (new StatusBundle).fromBits(0.U))
+
+
+  //////// Bit level ////////
+
+  io.port.scl.out := false.B                           // i2c clock line output
+  io.port.sda.out := false.B                           // i2c data line output
+
+  // filter SCL and SDA signals; (attempt to) remove glitches
+  val filterCnt = Reg(init = UInt(0, 14.W))
+  when ( !control.coreEn ) {
+    filterCnt := 0.U
+  } .elsewhen (~(filterCnt.orR)) {
+    filterCnt := Cat(prescaler.hi, prescaler.lo) >> 2  //16x I2C bus frequency
+  } .otherwise {
+    filterCnt := filterCnt - 1.U
+  }
+
+  val fSCL      = Reg(init = UInt(0x7, 3.W))
+  val fSDA      = Reg(init = UInt(0x7, 3.W))
+  when (~(filterCnt.orR)) {
+    fSCL := Cat(fSCL, io.port.scl.in)
+    fSDA := Cat(fSDA, io.port.sda.in)
+  }
+
+  val sSCL      = Reg(init = Bool(true), next = (new Majority(fSCL.toBools.toSet)).out)
```

Comment:
I thought that Bool(true) and true.B are exactly equivalent. But you're using both styles here, just suggesting that you be consistent.